iocage_lib/ioc_plugin.py

```diff
@@ -1246,29 +1236,69 @@ def __fetch_release__(self, release):
         fetch_args = {'release': release, 'eol': False}
         iocage_lib.iocage.IOCage(silent=self.silent).fetch(**fetch_args)
 
-    def __clone_repo(self, repo_url, destination):
+    @staticmethod
+    def _verify_git_repo(repo_url, destination):
+        verified = False
+        with contextlib.suppress(
+            git.exc.InvalidGitRepositoryError,
+            git.exc.NoSuchPathError,
+            AttributeError,
+        ):
+            repo = git.Repo(destination)
+            verified = any(u == repo_url for u in repo.remotes.origin.urls)
+
+        return verified
+
+    @staticmethod
+    def _clone_repo(ref, repo_url, destination, depth=None, callback=None):
         """
         This is to replicate the functionality of cloning/pulling a repo
         """
+        branch = ref
         try:
-            with open('/dev/null', 'wb') as devnull:
-                porcelain.pull(destination, repo_url, errstream=devnull)
-                repo = porcelain.open_repo(destination)
-        except dulwich.errors.NotGitRepository:
-            with open('/dev/null', 'wb') as devnull:
-                repo = porcelain.clone(
-                    repo_url, destination, errstream=devnull
-                )
-
-        remote_refs = porcelain.fetch(repo, repo_url)
-        ref = f'refs/heads/{self.branch}'.encode()
+            if not IOCPlugin._verify_git_repo(repo_url, destination):
+                raise git.exc.InvalidGitRepositoryError()
+
+            # "Pull"
+            repo = git.Repo(destination)
+            origin = repo.remotes.origin
+            ref = 'master' if f'origin/{ref}' not in repo.refs else ref
+            for command in [
+                ['checkout', ref],
+                ['pull']
+            ]:
+                cp = su.Popen(['git', '-C', destination] + command, stderr=su.DEVNULL, stdout=su.DEVNULL)
+                cp.communicate()
+                if cp.returncode:
+                    raise su.CalledProcessError(cp.returncode, cp.args)
+        except (
+            su.CalledProcessError,
+            git.exc.InvalidGitRepositoryError,
+            git.exc.NoSuchPathError
+        ):
+            # Clone
+            shutil.rmtree(destination, ignore_errors=True)
```

Comment:
rmtree always make me twitchy. Is there some verification we are not deleting a dir we shouldnt?